apache/netbeans
# use SMO service for class name and SHA1 queries.

service https://github.com/apache/maven-indexer/tree/master/search-backend-smo (search maven org)

 - the remote index does not contain class name information anymore
 - sha1 queries are rarely used and contribute to index size, which makes them a good candidate for SMO too, this will also allow to progress with #5922

![dependency-search-ui](https://github.com/apache/netbeans/assets/114367/056be4ac-e4f0-458d-9371-cc6b5b3ded2d)

SearchDependencyUI window update.
    
  - 'local search' checkbox to limit search to local repos
  - move search-in-progress marker to top of the list
  - other minor fixes (labels/resizing etc)
  - code renovation / deprecation fixes


sha1 query latency is fairly usable, here is the nbm wizard trying to resolve dependencies:
```
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=8, pageOffset=0}, query=sha1:afb2ce3d80d31a18a9d2ad532312e93a1fc8ad05} finished in 553 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=8, pageOffset=0}, query=sha1:ea59da7199e90dd8b1915ec5fa346b0fc0445235} finished in 124 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=8, pageOffset=0}, query=sha1:fe1f44c2917d776fc66c384b6314442a483d8801} finished in 150 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=8, pageOffset=0}, query=sha1:3392f14ad7125e85e46bcc98dfb7c323afef7f89} finished in 123 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=8, pageOffset=0}, query=sha1:e1a6e400b9cd9d0b42c03d368eb34c6d0e29a821} finished in 120 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=8, pageOffset=0}, query=sha1:4ea5950a57c2cde55d0281f61c3a16e66b0abcbd} finished in 123 ms
```

classname search can result in really bad latency, here a search for `StringUtils` which collects 10 pages (which I set as maximum for now):
```
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=0}, query=cn:StringUtils} finished in 9,222 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=1}, query=cn:StringUtils} finished in 4,413 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=2}, query=cn:StringUtils} finished in 6,406 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=3}, query=cn:StringUtils} finished in 6,725 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=4}, query=cn:StringUtils} finished in 5,083 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=5}, query=cn:StringUtils} finished in 854 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=6}, query=cn:StringUtils} finished in 5,426 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=7}, query=cn:StringUtils} finished in 7,202 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=8}, query=cn:StringUtils} finished in 6,863 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=9}, query=cn:StringUtils} finished in 2,110 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=10}, query=cn:StringUtils} finished in 3,981 ms
```
FQ class names are not much better:
```
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=0}, query=fqcn:dev.mbien.slf2jfr.JFRLogger} finished in 1,650 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=0}, query=fqcn:org.openide.awt.ActionID} finished in 4,446 ms
INFO [org.netbeans.modules.maven.indexer.SMORequestResult]: SMO SearchRequest{paging=Paging{pageSize=128, pageOffset=0}, query=fqcn:org.openide.util.NbBundle.Messages} finished in 2,355 ms
```


how to test:
open/create maven project, type some class name, use the hint to do a maven search. Results should contain both central and local entries. Should work for simple class names and fully qualified class names.


I am not sure how useful the sha1 query actually is, since features like attach src/doc do seem to work even with indexing disabled since it uses GAV. The local repo index does also contain hash information which means that as soon the project is primed, the local index could be used for it.
https://github.com/search?q=repo%3Aapache%2Fnetbeans+findBySHA1&type=code

please note that the `Dependencies -> Add Dependency...` window does use a different code path (generic `find` query with arbitrary fields). Going to update that one some time later.
